fix: GH-9103 Use DateTimePickerFormatterService in Replenishment Order

saad-mo opened the pull request:
closes #9103

- Moved `DatePickerFormatterService` and `DateTimePickerFormatterService` to the core to make it easier for other features to use these services 
- Used `DateTimePickerFormatterService` in `OccReplenishmentOrderFormSerializer` because we need accurate time in the string. I found that the order was off by a day when the time was being zeroed out (00:00:00). 
- Furthermore, the date picker returns a string with the current time only if the replenishment date is not changed. If the date is changed, the date picker does not return the current time. `OccReplenishmentOrderFormSerializer` has been updated to handle this `getCurrentLocalTime` function. 
- Finally, we are using `toModel` method from `DateTimePickerFormatterService` to append an accurate timezone to the date string. 


OLD PR: https://github.com/SAP/spartacus/pull/9234

Commit:
Use DatePickerFormatterService in Replenishment Order

Commit:
Remove test focus

Commit:
Export date-picker-formatter.service

Commit:
Merge branch 'develop' into feature/GH-9103

Commit:
Update imports to use @spartacus/core instead of relative paths

hackergil committed:
Merge branch 'develop' into feature/GH-9103

Commit:
Move date-time-picker-formatter service to core

Commit:
add toModelWithTime method to date-time-picker-formatter service + clean up

Commit:
Fix/Update unit tests

Commit:
Merge branch 'develop' into feature/GH-9103

Commit:
Merge branch 'develop' into feature/GH-9103

Commit:
Revert to using existing toModel method

Commit:
Update test to use mockservice and dynamic timezone

Commit:
Merge branch 'develop' into feature/GH-9103

Commit:
Update convertDate fuction to support date string with no time

Commit:
Merge branch 'develop' into feature/GH-9103

Commit:
update comments

bgambocjaviniar committed:
Merge branch 'develop' into feature/GH-9103

bgambocjaviniar commented on projects/core/src/occ/adapters/checkout/converters/occ-replenishment-order-form-serializer.ts:
```diff
@@ -27,10 +30,31 @@ export class OccReplenishmentOrderFormSerializer
+    const index = date.lastIndexOf(':');
+    let modelDate;
+    if (index > 0) {
+      modelDate = date.slice(0, index);
```
When index is bigger than 0, then that means a `:` is found. This case covers the issue, where the user did not choose a date from the datepicker. However, this logic is wrong, where you target the last index of a colon because the date string that was provided contains the hours and mins already which is set to UTC, therefore the hours being 4 hours ahead. 

For example, the default value that came contains `2011-10-05T14:48:00.000Z`, but 14:48 is the wrong datetime as the current eastern time would be 10:48, so instead of stripping the string from ':', we would strip the string from `T`.

All in all, I have provided a snippet and potential solution in the following link https://github.com/SAP/spartacus/pull/9457#discussion_r510454392

bgambocjaviniar commented on projects/core/src/occ/adapters/checkout/converters/occ-replenishment-order-form-serializer.ts:
```diff
@@ -27,10 +30,31 @@ export class OccReplenishmentOrderFormSerializer
+    if (index > 0) {
+      modelDate = date.slice(0, index);
+    } else {
+      modelDate = `${date}T${this.getCurrentLocalTime()}`;
```
👍

hackergil replied:
I got rid of this. Service should just get a date with no time

bgambocjaviniar commented on projects/core/src/occ/adapters/checkout/converters/occ-replenishment-order-form-serializer.spec.ts:
```diff
@@ -1,29 +1,55 @@
-  it('should convert the date string of schedule replenishment form to ISO 8601 format', () => {
-    const mockDate = '1994-01-11';
-    const ISODateFormat = mockDate + 'T00:00:00Z';
+  it('should convert the date string with time', () => {
```
add a test for `getCurrentLocalTime` please

hackergil replied:
I think there's no need anymore.

bgambocjaviniar commented on projects/core/src/occ/adapters/checkout/converters/occ-replenishment-order-form-serializer.spec.ts:
```diff
@@ -1,29 +1,55 @@
-    const mockDate = '1994-01-11';
-    const ISODateFormat = mockDate + 'T00:00:00Z';
+  it('should convert the date string with time', () => {
+    const mockDateWithTime = '1994-01-11:01:02:03.456Z';
```
As mentioned in the component, where I inserted comments, the date by default without choosing a date from the datepicker comes in utc time already, so this mock date with time is wrong as it won't represent the case you're trying to test

bgambocjaviniar replied:
The format is wrong too, where the mock date has a `T` and not a `:`

I would fix to have some data consistency even if mocked for this special representation.

hackergil replied:
Fixed

bgambocjaviniar commented on projects/core/src/occ/adapters/checkout/converters/occ-replenishment-order-form-serializer.ts:
```diff
@@ -27,10 +30,31 @@ export class OccReplenishmentOrderFormSerializer
 
   /**
-   * Converts the date string to the Standard ISO 8601 format
+   * Converts the date string from a Standard ISO 860 format with or without time to valid OCC model format
```
I would not change the comment as the ' accepted occ model format' is literally a Standard ISO 8601 format

bgambocjaviniar replied:
Moreover, there's a typo, where you mentioned 860 and not 8601

bgambocjaviniar commented on projects/core/src/occ/adapters/checkout/converters/occ-replenishment-order-form-serializer.ts:
```diff
@@ -27,10 +30,31 @@ export class OccReplenishmentOrderFormSerializer
+      modelDate = `${date}T${this.getCurrentLocalTime()}`;
+    }
+
+    return this.dateTimePickerFormatterService.toModel(modelDate);
```
*Potential solution

I went ahead and look through a fix
Hopefully, my provided snippet will help with your problem.

```
private convertDate(date: string): string {
// native date will always give YYYY-MM-DD whether the default date
// was sent from the replenishment form or from the datepicker
    const nativeDate = this.datePickerFormatterService.toNative(date);

**// potential re-work or added value to the toModel function 
// either pass a param that will set the HH:MM:ss||00 when given,** 
// therefore needing to massage the logic in the toModel function of the DatePickerFormatter service. 
// Verify with Caine if that logic can or should be added to that service. 
//I believe it should as we're working with a datepicker and not a datetime-local. What do you think?
    const modelDate = this.datePickerFormatterService.toModel(
      nativeDate,
      false,
      this.getCurrentLocalTime ?
    );

 // modelDate would contain the value YYYY-MM-DDTHH:MM:SSZ or YYYY-MM-DDTHH:MM:SS+00:00 
 // YYYY-MM-DD = the nativeDate
 // HH:MM:SS can be given from the `getCurrentLocalTime` function which will most likely be moved to datepicker service
    return modelDate;
  }
```

After looking through the snippet above, 

please do the following

1) need to checkup with someone to check out the backend as I believe we don't have to set the offset, and we only needed to set the datetime HH:mm:ss in order to get the proper value back

bgambocjaviniar reviewed (changes requested):
Small quick question: Why was the PR necessary to re-create?

--------

I dug dipper and the fix for this date problem is just to provide the datetime hh:mm:00 with no offset of timezone, but with +00:00

Moreover, from my findings and additional testing, I believe there is no need to pass an offset of a timezone as that does not work well from the given response of a request as it will add/substract to the current time that was originally sent. (from here https://github.com/SAP/spartacus/pull/9234#discussion_r503218703)

What was sent as a request
![Screen Shot 2020-10-22 at 5 06 19 PM](https://user-images.githubusercontent.com/29099466/96929964-f85ec400-1488-11eb-9f3d-55b88fa71706.png)

What we receive as response
![Screen Shot 2020-10-22 at 5 06 30 PM](https://user-images.githubusercontent.com/29099466/96929994-03195900-1489-11eb-9ce8-05fff13bcadc.png)

As you can see, what was being sent is 17:06, but the response that was given due to the offset became 21:06. The offset added additional time to the original current datetime of the date.

I proposed a potential solution with a snippet in https://github.com/SAP/spartacus/pull/9457#discussion_r510454392

hackergil committed:
Merge branch 'develop' into feature/GH-9103

hackergil committed:
GH-9103 Checkpoint.
- Moving ISO date creation to replenishment form service.
- Updating replenishment order component accordingly
- Updating test data

hackergil committed:
GH-9103 Another checkpoint. Simplifying date generation for OCC

bgambocjaviniar commented on projects/storefrontlib/src/cms-components/checkout/services/checkout-replenishment-form-service.ts:
```diff
@@ -19,7 +19,7 @@ export class CheckoutReplenishmentFormService {
     numberOfWeeks: '1',
     recurrencePeriod: recurrencePeriod.DAILY,
-    replenishmentStartDate: new Date().toISOString(),
+    replenishmentStartDate: new Date().toISOString().split('T')[0],
```
👍  Better than changing it in the serializer

bgambocjaviniar reviewed (changes requested):
LGTM except the offset, we need to talk about this before approval
Thanks for picking it up Gil.

bgambocjaviniar committed:
Merge branch 'develop' into feature/GH-9103

hackergil committed:
Merge branch 'develop' into feature/GH-9103

hackergil committed:
Merge branch 'develop' into feature/GH-9103

hackergil committed:
Merge branch 'develop' into feature/GH-9103
